### ISSUE TITLE
Possible data structure corruption

### ISSUE THREAD

**Issue by pturschmid** [YanVugenfirer/kvm-guest-drivers-windows#65]
We saw a crash that seems to happen when pausing the NetKvm adapter. We recovered the following 2 stacks from the dump.

```
00 fffff802`4a6a3c08 fffff802`4903c1c2 nt!KeBugCheckEx 
01 fffff802`4a6a3c10 fffff802`49232e4d hal!HalBugCheckSystem+0x7e 
02 fffff802`4a6a3c50 fffff802`4903cfa1 nt!WheaReportHwError+0x22d 
03 fffff802`4a6a3cb0 fffff802`49256c20 hal!HalHandleNMI+0xfe 
04 fffff802`4a6a3ce0 fffff802`491cc7c2 nt!KiProcessNMI+0x150 
05 fffff802`4a6a3d30 fffff802`491cc636 nt!KxNmiInterrupt+0x82 
06 fffff802`4a6a3e70 fffff802`49101d6e nt!KiNmiInterrupt+0x176 
07 fffff802`4a698950 fffff800`57bdd768 nt!KxWaitForSpinLockAndAcquire+0x22 
08 fffff802`4a698980 fffff800`57bde585 netkvm!CNdisList<CNBL,CRawAccess,CNonCountingObject>::ForEachPrepareIf<<lambda_6e3a6aeb098631fcc79aaecc43ea142b>,<lambda_0fb1256e1da6df1a035ab89e05289ff7>,<lambda_2e9bd86eeab3a657911ddcd0e23da5de> >+0x4 
09 fffff802`4a6989c0 fffff800`57bd8061 netkvm!CNBL::ParseBuffers+0xbd 
0a fffff802`4a698a70 fffff800`57be351c netkvm!ParaNdis_CleanupContext+0x71 
0b fffff802`4a698aa0 fffff800`55ea8e12 netkvm!ParaNdis6_SendPauseRestart+0x50 
0c fffff802`4a698b00 fffff802`490b2f10 NDIS!ndisInterruptDpc+0x1a3 
0d fffff802`4a698be0 fffff802`490b2257 nt!KiExecuteAllDpcs+0x1b0 
0e fffff802`4a698d30 fffff802`491c63d5 nt!KiRetireDpcList+0xd7
```

```
00 ffffd001`5abf8fe0 fffff802`49101d32 nt!KxWaitForSpinLockAndAcquire+0x1f 
01 ffffd001`5abf9010 fffff800`57bde711 nt!KeAcquireSpinLockRaiseToDpc+0x32 
02 ffffd001`5abf9040 fffff800`57bdeecc netkvm!CNBL::ParseOffloads+0xc5 
03 ffffd001`5abf9070 fffff802`4900ba37 netkvm!CParaNdisTX::SendMapped+0x3c 
04 ffffd001`5abf90a0 fffff800`55eade82 hal!HalBuildScatterGatherListV2+0x207 
05 ffffd001`5abf9140 fffff800`57bdf09a NDIS!NdisMAllocateNetBufferSGList+0x1a2 
06 (Inline Function) --------`-------- netkvm!RtlFailFast+0x2 
07 (Inline Function) --------`-------- netkvm!FatalListEntryError+0x2 
08 (Inline Function) --------`-------- netkvm!InsertHeadList+0x10c 
09 (Inline Function) --------`-------- netkvm!CNdisList<CNBL,CRawAccess,CNonCountingObject>::Push+0x120 
0a ffffd001`5abf9200 fffff800`57bdda02 netkvm!CParaNdisTX::SendMapped+0x20a 
0b (Inline Function) --------`-------- netkvm!CNBL::ParseOffloads::__l39::<lambda_45c6048e0eee80e44000724f00745ebe>::operator()+0x18 
0c ffffd001`5abf9240 fffff800`57bdf58f netkvm!CNBL::ParseCSO<<lambda_07bb9e4d8a255eab68dc9a6db8158716>,<lambda_8edd56c1927263d1d00c45709877ef5c>,<lambda_45c6048e0eee80e44000724f00745ebe> >+0x2a 
0d ffffd001`5abf9270 fffff800`57bdf1a5 netkvm!CTXVirtQueue::Create+0x43 
0e ffffd001`5abf92a0 fffff800`57be2dce netkvm!CNB::SetupLSO+0xc1 
```

I think the cause is that m_SendList uses the CRawAccess strategy and there's a race between Remove_LockLess in RemoveAllNonWaitingNBLs and Push in PushMappedNBL. This causes the FatalListEntryError I believe. Using RemoveHeadList instead of RemoveEntryList may be threadsafe. Otherwise I was thinking the m_SendList may need a new barrier to prevent adding new entries when pausing the adapter.

**Comment by YanVugenfirer** [comment]
Thank you for the report and analysis. We are reworking the "Pause" behaviour right now and will address all the issues related to it.

**Comment by pturschmid** [comment]
By any chance do you have an estimate of when you'll rework the pause behavior? Alternately can you provide a description of what the fix may look like? I want to try a fix ahead of time if possible to see if it makes our repro go away. :-) Thanks!

**Comment by YanVugenfirer** [comment]
We are testing the code now with HCK (MS certification tool). I think
during next week we will have a stable code for sharing.

Best regards,
Yan.


On Tue, May 24, 2016 at 9:56 PM, pturschmid <notifications@github.com>
wrote:

> By any chance do you have an estimate of when you'll rework the pause
> behavior? Alternately can you provide a description of what the fix may
> look like? I want to try a fix ahead of time if possible to see if it makes
> our repro go away. :-) Thanks!
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/YanVugenfirer/kvm-guest-drivers-windows/issues/65#issuecomment-221364484>
>



-- 
-----------------------------------------------------------------------
Daynix Computing LTD
Yan Vugenfirer, CEO
Email: yan@daynix.com
Phone (Israel): +972-54-4758084
Phone (USA): +1-7204776716
Phone (UK): +44-2070482938
Web: www.daynix.com

**Comment by pturschmid** [comment]
Thanks Yan for the super quick turnaround! :-)

**Comment by YanVugenfirer** [comment]
Please try the following build: https://dl.dropboxusercontent.com/u/97739133/virtio-win-prewhql-0.1-118.zip

**Comment by pturschmid** [comment]
Thanks Yan. Will try it out ASAP! :-)

**Comment by pturschmid** [comment]
Good news! Fixed the bug perfectly! Thanks!

**Comment by YanVugenfirer** [comment]
Thank you for the feedback!

**Comment by YanVugenfirer** [comment]
There were two patch sets that addressed the issue: bacc1c242cfc747832b5211ccfcb317a7904c509-911bad680ee29619fa521e0d87b3e1b6c7f2dd3f and e1de11fd364a662d097970301976a92efb715740-7ef084b8f739e524bc690bda876c58ba955646c7
